aquadoggo/src/replication/service.rs

```diff
@@ -0,0 +1,398 @@
+            .handle_message(&PeerConnectionId(peer_id, Some(connection_id)), &message)
+            .await
+        {
+            Ok(result) => {
+                for message in result.messages {
+                    self.send_service_message(ServiceMessage::SentReplicationMessage(
+                        peer_id,
+                        connection_id,
+                        message,
+                    ));
+                }
+
+                if result.is_done {
+                    self.on_replication_finished(peer_id, connection_id, session_id)
+                        .await;
+                }
+            }
+            Err(err) => {
+                self.on_replication_error(peer_id, connection_id, session_id, err)
+                    .await;
+            }
+        }
+    }
+
+    async fn on_replication_finished(
+        &mut self,
+        peer_id: PeerId,
+        _connection_id: ConnectionId,
+        _session_id: SessionId,
+    ) {
+        info!("Finished replication with peer {}", peer_id);
+        match self.connections.get_mut(&peer_id) {
+            Some(peer) => {
+                peer.status.successful_count += 1;
+            }
+            None => {
+                panic!("Tried to access unknown peer");
+            }
+        }
+    }
+
+    async fn on_replication_error(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        session_id: SessionId,
+        error: ReplicationError,
+    ) {
+        warn!("Replication with peer {} failed: {}", peer_id, error);
+
+        match self.connections.get_mut(&peer_id) {
+            Some(peer) => {
+                peer.status.failed_count += 1;
+            }
+            None => {
+                panic!("Tried to access unknown peer");
+            }
+        }
+
+        match error {
+            ReplicationError::StrategyFailed(_) | ReplicationError::Validation(_) => {
+                let peer_connection_id = PeerConnectionId(peer_id, Some(connection_id));
+                self.sync_manager
+                    .remove_session(&peer_connection_id, &session_id);
+            }
+            _ => (), // Don't try and close the session on other errors as it should not have been initiated
+        }
+    }
+
+    async fn handle_service_message(&mut self, message: ServiceMessage) {
+        match message {
+            ServiceMessage::ConnectionEstablished(peer_id, connection_id) => {
+                self.on_connection_established(peer_id, connection_id).await;
+            }
+            ServiceMessage::ConnectionClosed(peer_id, connection_id) => {
+                self.on_connection_closed(peer_id, connection_id).await;
+            }
+            ServiceMessage::ReceivedReplicationMessage(peer_id, connection_id, message) => {
+                self.on_replication_message(peer_id, message, connection_id)
+                    .await;
+            }
+            ServiceMessage::InitiateReplication => {
```

Comment:
This comes from the scheduler and is now the only moment replication gets triggered.